dymos/transcriptions/transcription_base.py

```diff
@@ -132,6 +132,112 @@ def setup_controls(self, phase):
             phase.add_subsystem('control_group',
                                 subsys=control_group)
 
+    def _configure_state_introspection(self, state_name, options, phase):
+        """
+        Modifies state options in-place, automatically determining 'targets', 'units', and 'shape'
+        if necessary.
+        The precedence rules for the state shape and units are as follows:
+        1. If the user has specified units and shape in the state options, use those.
+        2a. If the user has not specified shape, and targets exist, then pull the shape from the targets.
+        2b. If the user has not specified shape and no targets exist, then pull the shape from the rate source.
+        2c. If shape cannot be inferred, assume (1,)
+        3a. If the user has not specified units, first try to pull units from a target
+        3b. If there are no targets, pull units from the rate source and multiply by time units.
+        For units, specifically, option 3b can lead to ugly behavior.  If the state rate has units
+        of 'm/s', then the state units will be inferred as 'm/s*s'.  While mathematically correct,
+        OpenMDAO does not currently simplify unit strings.
+        Parameters
+        ----------
+        state_name : str
+            The name of the state variable of interest.
+        state_options: OptionsDictionary
+            The options dictionary for the state variable of interest.
+        phase
+            The phase associated with the transcription.
+        """
+        time_units = phase.time_options['units']
+        user_targets = options['targets']
+        user_units = options['units']
+        user_shape = options['shape']
+
+        need_units = user_units is _unspecified
+        need_shape = user_shape in {None, _unspecified}
+
+        ode = phase._get_subsystem(self._rhs_source)
+
+        # Automatically determine targets of state if left _unspecified
+        if user_targets is _unspecified:
+            from dymos.utils.introspection import get_targets
+            options['targets'] = get_targets(ode, state_name, user_targets)
+
+        # 1. No introspection necessary
+        if not(need_shape or need_units):
+            return
+
+        # 2. Attempt target introspection
+        if options['targets']:
+            try:
+                from dymos.utils.introspection import get_state_target_metadata
+                tgt_shape, tgt_units = get_state_target_metadata(ode, state_name, options['targets'],
+                                                                 options['units'], options['shape'])
+                options['shape'] = tgt_shape
+                options['units'] = tgt_units
+                return
+            except ValueError:
+                pass
+
+        # 3. Attempt rate-source introspection
+        rate_src = options['rate_source']
+        rate_src_type = phase.classify_var(rate_src)
+
+        if rate_src_type in ['time', 'time_phase']:
```

Comment:
Do we do all this somewhere else too?  If so, maybe put it into a new function `get_state_rate_src_metadata` in the introspection module.

Reply:
That would likely be a good idea.  A significant refactoring is probably in order, but maybe we should wait until after 1.0 and we drop RungeKutta to reduce the pain from it.